iree/compiler/Codegen/LLVMGPU/LLVMGPUTileAndDistribute.cpp

```diff
@@ -235,6 +232,23 @@ struct LLVMGPUTileAndDistributePass
       OwningRewritePatternList promotionPatterns(&getContext());
       populatePromotionPatterns(context, promotionPatterns);
       (void)applyPatternsAndFoldGreedily(funcOp, std::move(promotionPatterns));
+      // Insert barriers before and after copies to workgroup memory and skip
+      // insert barriers between back to back copy to workgroup memory.
+      OpBuilder builder(&getContext());
+      funcOp.walk([&builder](linalg::CopyOp copyOp) {
+        if (hasMarker(copyOp, getCopyToWorkgroupMemoryMarker())) {
+          Operation *prevOp = copyOp->getPrevNode();
```

Comment:
Isnt this a bit fragile? There might be other non-data movement related operations here.

I dont have a better solution, but we probably want to start thinking about using tensors at this level. That will make it easier to reason about where barriers go.
If you want to stay in buffers, at least we should use something like `DependenceGraph` (But that is only a little less fragile).

Reply:
It's fragile in the sense that it may insert extra barriers but it shouldn't cause wrong code to be emitted.
I agree we should start thinking about it on the tensor level, although I don't think the solution is going to be trivial there either and the semantic of barriers at the tensor level is going to be an interesting problem.